### PR TITLE
[inductor][cpu] Fix accuracy error in BMM benchmarking for input weight with offset

### LINKED ISSUE #143770: [inductor][cpu] Accuracy failure on bmm max_autotune for offset input weights
### 🐛 Describe the bug

Accuracy error is occurring for BMM max_autotune code when input weights have an offset. Issue is not reproducible on main due to #143102 but after #143141 lands, this issue shows up. Found testing torchbench `sam` model with `--amp`.

Here is a sample test to reproduce (could be added to `test/inductor/test_cpu_select_algorithm.py`):
```python
    @patches
    @torch.no_grad
    @unittest.skipIf(not TEST_MKL, "Test requires MKL")
    @dtypes(torch.bfloat16)
    def test_bmm_5d(self, dtype):
        class M(torch.nn.Module):
            def __init__(self):
                super().__init__()

            def forward(self, x, w):
                return x @ w[2]

        counters.clear()
        x = torch.randn(400, 196, 196).to(dtype=dtype)
        w = torch.randn(3, 400, 196, 80).to(dtype=dtype)
        mod = M().to(dtype=dtype).eval()
        with verify(dtype) as (atol, rtol):
            self.common(mod, (x, w), atol=atol, rtol=rtol)
        self.assertEqual(counters["inductor"]["select_algorithm_autotune"], 1)
```

The error seems to be related to taking the `as_strided` tensor in `normalize_shapes` in `cpp_gemm_template.py`, but more investigation is needed.

### Versions

Seen in main after cherry-picking from #143141

### PR DESCRIPTION
**Pull request by frost-intel**
Fixes #143770 

When an input weight tensor has an offset (i.e. is a slice of another larger tensor at non-zero dim) the test/benchmarking process was changing the benchmarking argument to be only the one slice instead of the entire tensor. This resulted in an accuracy error and potentially a crash if in `VERIFY` mode in `select_algorithm.py`.

As a solution, we check if the input weight is a slice of a larger node, and if so, we use the larger node for the call to `as_strided` when preprocessing the benchmarking arguments.

* Why wasn't this happening before with GEMM?
  - Since current GEMM code only supports constant weights, the blocking/packing process changed the input weight tensor so no offset was used. This is not the case for BMM.

The new UT here tests both the BMM and GEMM cases, where the GEMM input is a slice and a constant weight, and the BMM input is not constant.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

### CODE REVIEW

File: torch/_inductor/codegen/cpp_bmm_template.py [hunk 1]
```diff
@@ -131,6 +131,27 @@ def check_if_block_weight(W, micro_gemm):
...
+                new_layout = node.layout
+                return (
+                    isinstance(node.data, ir.StorageBox)
+                    and old_layout.size[-len(new_layout.size) :] == new_layout.size
+                    and len(old_layout.size) > len(new_layout.size)
+                    and new_layout.is_contiguous()
+                )
+            return False
+
+        if is_slice(node):
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp_bmm_template.py:148]: can we just unwrap the view and get the unwrapped layout? also it is not specific to bmm but nonconst weight?

Comment (leslie-fang-intel) [reply torch/_inductor/codegen/cpp_bmm_template.py:148]: > also it is not specific to bmm but nonconst weight?

I think for GEMM Template with nonconst weight, we didn't meet this kind of issue because it requires contiguous layout (we blocked the mat2 into block format and make the blocked mat2 contiguous). It will introduce a copy before the GEMM Template.

https://github.com/pytorch/pytorch/blob/bc576355a2672ac1a7fb2a2c50d8e714df367a16/torch/_inductor/codegen/cpp_gemm_template.py#L1092-L1093

Comment (frost-intel) [reply torch/_inductor/codegen/cpp_bmm_template.py:148]: We don't always want the unwrapped layout, only if the weights returned from the `as_strided` call below has a different address. 
https://github.com/pytorch/pytorch/blob/bc576355a2672ac1a7fb2a2c50d8e714df367a16/torch/_inductor/codegen/cpp_gemm_template.py#L859-L863

For example, if the view is a reshape op combining dims, we don't want to unwrap.

Yes, this is not BMM specific, but is specific to non-constant weight. I could move the `is_slice` logic into `cpp_gemm_template.py` and add a check for non-constant weight?

File: torch/_inductor/codegen/cpp_gemm_template.py [hunk 2]
```diff
@@ -896,6 +906,17 @@ def normalize_shapes(inputs, layout_or_out):
...
         block_weights = cls.check_if_block_weight(new_inputs[1], micro_gemm)
 
+        # Note that the layout of MKLDNN Tensor is with the wrong stride
+        if is_slice(original_weight_node) and not block_weights:
+            # If weight is a slice of a larger tensor, we need to use the layout of the whole tensor
+            # instead of just the slice because the GEMM template expects the whole tensor.
+            view_layout = original_weight_node.data.get_layout()
+        else:
+            view_layout = original_weight_node.layout
+        view_size = view_layout.size
```
Comment (leslie-fang-intel) [inline torch/_inductor/codegen/cpp_gemm_template.py:916]: If my understanding is right, for the case we fixed here, it seems we can directly use the whole tensor without view it back for the benchmark.

Comment (frost-intel) [reply torch/_inductor/codegen/cpp_gemm_template.py]: You're right, I added code to skip the view if we need to use the whole tensor.

File: test/inductor/test_cpu_select_algorithm.py [hunk 3]
```diff
@@ -2000,6 +2000,48 @@ def forward(self, x, w):
         self.assertEqual(counters["inductor"]["select_algorithm_autotune"], 1)
         self.assertEqual(counters["inductor"]["cpp_epilogue_fusion_counter"], 1)
 
+    @patches
+    @torch.no_grad
+    @unittest.skipIf(not TEST_MKL, "Test requires MKL")
+    @dtypes(torch.float)
+    def test_weight_offset(self, dtype):
```
[the remark this answers is not in the document]
Comment (frost-intel) [reply test/inductor/test_cpu_select_algorithm.py]: Added test for symbolic size.

File: torch/_inductor/codegen/cpp_gemm_template.py [hunk 4]
```diff
@@ -299,6 +299,18 @@ def get_padded_n(n, block_n):
     return (n + block_n - 1) // block_n * block_n
 
 
+def is_slice(node: ir.IRNode):
+    if isinstance(node, ir.ReinterpretView):
+        old_layout = node.data.get_layout()
+        new_layout = node.layout
+        return (
+            isinstance(node.data, ir.StorageBox)
+            and old_layout.size[-len(new_layout.size) :] == new_layout.size
```
[the remark this answers is not in the document]
Comment (frost-intel) [reply torch/_inductor/codegen/cpp_gemm_template.py]: New code evaluates any symbolic sizes, then checks that the new layout is smaller or equal in all dimensions, not just a slice in the first dim.

File: torch/_inductor/codegen/cpp_gemm_template.py [hunk 5]
```diff
@@ -896,6 +922,11 @@ def normalize_shapes(inputs, layout_or_out):
         assert micro_gemm is not None
         block_weights = cls.check_if_block_weight(new_inputs[1], micro_gemm)
 
+        # If weight is a slice of a larger tensor, we need the entire tensor and shouldn't unwrap it
+        # because the GEMM template expects the whole tensor.
+        if not block_weights and is_slice(original_weight_node):
```
Comment (leslie-fang-intel) [inline torch/_inductor/codegen/cpp_gemm_template.py:927]: Please also add the comment why `should_unwrap_weight` relates to `block_weights`.
